feat: Add vw as python module command

mathcass opened the pull request:
The main intention here is to add a new command that would allow someone to install the Python package and use it as if it were the CLI. There's been a lot of work to ensure that installing via Python works across various environments. I figured this might be an easy way to translate that easy installation into the CLI interface as well. 

With this change, the experience would look a lot like, 

```sh
python -m pip install vowpalwabbit
python -m vw [normal command-line arguments]
```

As a brief demo of this, I put together a change in the dependencyparsing demo to show that this works, https://github.com/VowpalWabbit/vowpal_wabbit/compare/ecadd8d8b622a...fb9a501

```sh
docker build . -t vowpal-wabbit:module
docker run -it vowpal-wabbit:module sh -c "cd /vowpal_wabbit/demo/dependencyparsing/ && make dep.perf"
```

Would this be a helpful addition to the Python package? Any issues with including it? Any suggestions to test the expected behavior?

lalo commented:
This seems relevant
https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html

lalo commented:
FYI: We also have this docker image with the executable https://hub.docker.com/r/vowpalwabbit/vw-rel-alpine

but this idea is also welcomed. Let's figure out if there's functionality with setup.py

mathcass commented:
Thanks @lalo. I'm glad it would be welcomed. 

If I'm reading correctly, it sounds like you'd prefer that when a user runs `pip install vowpalwabbit` that they also get a `vw` command available to them from the installed Python package? 

Doing so would be a simple fix, but I avoided doing so because it didn't seem like a good practice to have a Python package mask a native executable that's potentially available. I tend to like the way [spaCy](https://spacy.io/usage) has implemented their sets of module CLI commands. And I wanted usage to explicitly reflect that this wasn't Vowpal Wabbit running as a binary executable, it was running as a script from the Python module. 

I can take the next steps to clean this up. Where I'd like some feedback is:

* Would you prefer usage to look like?
    1. Install via `python -m pip install vowpalwabbit`
    2. Run via:
        * Both: `vw` and `python -m vw`
        * Only: `vw`
        * Only: `python -m vw` - IE, was any of what I said above compelling enough to keep this around.

mathcass commented:
@lalo I proceeded with the second option "Only: `vw`" since that's what you mentioned in your original message. Let me know if you think otherwise. 

I also rebased & squashed all the changes (old changes still exist on my fork as separate branches).

lalo commented:
Sorry for not seeing this earlier, you raise valid points. Let me think about it. also @jackgerrits what are your thoughts here?

lalo commented:
I would lean towards installing a command named pyvw to avoid masking vw executable (also giving the convenience of having both at the same time). We can update documentation to explain this extra command available that mimics the executable.

jackgerrits commented:
I agree on the shadowing of the native binary. So I wonder if the proper thing here is to actually bundle the shipped binary? CMake does this: https://pypi.org/project/cmake/

mathcass commented:
> I agree on the shadowing of the native binary. So I wonder if the proper thing here is to actually bundle the shipped binary? CMake does this: https://pypi.org/project/cmake/

I think that would certainly make it easier to install for some folks. 

I tracked down (to the codepath) where they do this and it looks like it's here, https://github.com/scikit-build/cmake-python-distributions/blob/master/src/cmake/__init__.py#L42-L55

There's some pros & cons to either approach, though. There's still some work left to do if we'd want to ship the binary. It seems like you'd want do set up the package to publish the Python package for various systems. 

The simplicity of the pure Python version (this PR) is that it's making use of the convenience that folks have already put together tooling for building the C++ extension that is bundle with each Python package (❓ - Is it the same extension used for every platform? If so, then that's very convenient.)

If we want to proceed with shipping the binary, is there someone who knows more about the build system that could help with some pointers for hooking that into the Python installation?

jackgerrits commented:
One large con of shipping the binary is that it will effectively double the size of the Python package. Using the method you suggested is close enough to the behavior of the standard command line tool to be equivalent and saves a lot on space and complexity of packaging.

Same C++ extension for each platform yes.

Because of size and packaging complexity I think I'm for your original approach. I do wonder if installing as a `vw` module is counter intuitive. Why not make it available under `python -m vowpalwabbit` and if a user wishes to make it easier to run they can add an alias? Or is that enough friction to be an issue?

mathcass commented:
> Because of size and packaging complexity I think I'm for your original approach. I do wonder if installing as a `vw` module is counter intuitive. Why not make it available under `python -m vowpalwabbit` and if a user wishes to make it easier to run they can add an alias? Or is that enough friction to be an issue?

I originally thought that a `vw` module would be nice since it is obvious that running `vw` and running `python -m vw` were equivalent. But, I don't think would generally be a problem with running as `python -m vowpalwabbit`. You might argue that it's kind of a "branding" question. Do you want folks associating the string "vowpalwabbit" with invocation of this tool or preserve "vw" at all? As you probably see from the diff, making this `python -m vowpalwabbit` is even easier than what I've done in the PR (indeed, making `python -m vw` work involves effectively making a "shim" module). 

Before I proceed with the next update, sounds like we want, 

* Change: `python -m vw` -> `python -m vowpalwabbit`
* Remove: Remove the `pyvw` console script

Does that sound right?

mathcass commented:
Another ❓  -- With respect to ensuring that running as a module command works across different uses, do you have any tips for verifying that? 

It looked like one method might be to (for the purposes of testing), patch [`find_vw_binary`](https://github.com/mathcass/vowpal_wabbit/blob/python-module-script/test/run_tests.py#L443) so that it returns the string `python -m vowpalwabbit` and then run the tests. Is there an easier (or more straightforward) way?

lalo commented:
Here:

https://github.com/VowpalWabbit/vowpal_wabbit/blob/aa45c12259cc65185ec7e5b2ed0db39dae58c8e3/.scripts/linux/test.sh#L10

with this argument (--vw_bin_path)

https://github.com/VowpalWabbit/vowpal_wabbit/blob/aa45c12259cc65185ec7e5b2ed0db39dae58c8e3/test/run_tests.py#L644

Not sure if its going to be compatible with "python -m vowpalwabbit" though, what are your thoughts @jackgerrits

mathcass commented:
Thanks @lalo ! I think I got testing figured out. It doesn't feel ideal but it got it done. 

Here are the full test results (with links to the individual ones for native binary vs Python module), https://gist.github.com/mathcass/25022d531c72700cede081bb21a75642

The conclusion is Native Binary:

```
# Success: 249
# Fail: 60
# Skip: 31
```

and Python Module:

```
# Success: 241
# Fail: 67
# Skip: 32
```

👀 inspecting some of the diffs make them look like differences in numeric precision for various runtime values. For instance, Test 3 fails because `Diff not OK, Lines differ - ref vs output: 'weighted label sum = 526.517586' vs 'weighted label sum = 526.517528'`

I'm not entirely sure *why* they would differ, though, if they're using the same base linked library. (Unless there's something happening at the layer of Python loading it and running it.)

---

Where that leaves this is:

* I trimmed this down to just `python -m vowpalwabbit` rather than `python -m vw` or having a Python installed entrypoint `vw`
* I've added a hook into the tests to allow testing the installed `python -m vowpalwabbit` module command
* I added a command to run the new command in CI (hopefully it's set up correctly)

jackgerrits commented:
Thanks for your work @mathcass! I think we want to put the testing in the respective build python wheels workflow for each platform. For example Linux is https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/.github/workflows/build_python_wheels.yml

mathcass commented:
> Thanks for your work @mathcass! I think we want to put the testing in the respective build python wheels workflow for each platform. For example Linux is https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/.github/workflows/build_python_wheels.yml

Sorry for the radio silence @jackgerrits. Coming to look at this again and tried putting the test in https://github.com/VowpalWabbit/vowpal_wabbit/pull/3144/commits/dd1cb7125ae1a6ccb8fc36d29c3ff7683befad90

jackgerrits commented:
Thanks for picking this back up! Looks like it fails on finding `spanning_tree`. I think for this test run we should make it possible to skip those tests, that will require a change in `run_tests.py`. I'll see if I can take a look at that some time soon.

jackgerrits commented:
You should be able to resolve the test issue now by supplying `--skip_spanning_tree_tests` to `run_tests.py`

jackgerrits commented on python/vowpalwabbit/__main__.py:
```diff
@@ -0,0 +1,12 @@
+
+def main():
+    opts = sys.argv[1:]
+    pyvw.vw(" ".join(opts))
```
If the run_tests.py script works without it then I think we are good here

jackgerrits commented:
Just added the required float tolerance the same as other CI and fixed a conflict

mathcass commented:
@jackgerrits Thanks so much for handling those remaining pieces on precision and for driving the rest of this work. I'm excited to see this in the main package ❗ I appreciate your stewardship here.
